tensorboard/java/org/tensorflow/tensorboard/vulcanize/Vulcanize.java

```diff
@@ -707,6 +718,46 @@ private static boolean shouldIgnoreUri(String uri) {
     return ImmutableMultimap.copyOf(builder);
   }
 
+  // Combines all script tags into one and append it at the bottom of the
+  // document.
+  private static void combineScriptTags(Document document) {
+    Elements scripts = document.getElementsByTag("script");
+    final StringBuilder sources = new StringBuilder();
+    for (Element script : scripts) {
+      if (!script.attr("src").equals("")) continue;
+      sources.append(script.html()).append("\n");
+      script.remove();
+    }
+    // TensorBoard HTMLs tend to be extremely malformed and it can have multiple
+    // body tags.
+    Elements bodies = document.getElementsByTag("body");
+    Element lastBody = bodies.get(bodies.size() - 1);
+
+    Element scriptTag = new Element(Tag.valueOf("script"), "")
+        .appendChild(new DataNode(sources.toString(), ""));
+    lastBody.appendChild(scriptTag);
+  }
+
+  private static ArrayList<String> computeScriptShasum(Document document) {
+    ArrayList hashes = new ArrayList<String>();
+    for (Element script : document.getElementsByTag("script")) {
+      String hash = Hashing.sha256().hashString(script.html(), UTF_8).toString();
```

Comment:
Maybe add an assertion here that `src` is empty, because otherwise we
won’t actually set up the CSP properly, right?

Reply:
Thanks for pointing this out. I changed the logic a little now.

Since we have full filepath to all webfiles, we are actually able to get the source content of the script `src` so we now shasum that too.